### DISCUSSION TITLE
how to draw text utf-8 or std::wstring

### DISCUSSION THREAD

**Discussion by xjgong** [vsg-dev/VulkanSceneGraph#844] [Q&A] [answered]
Hello, I want to render text(utf-8),so i modified the vsgExamples\vsgtext.cpp and test.
1. When use vsgXchange to read Font(such as *.ttf in C:/Windows/Fonts/), the process always too slow. So I use vsgconvd.exe to convert the *.ttf to *.vsgt file first,then load vsgt file. Loading vsgt file is not slow,is this is the right way to use Font?
2. when I want to render Chinese "好"(\u597d),I found that there is no std::wstring defined with VSG_value. So I converted the Font file(ttf to vsgb),then try to use methods:
   text->text = vsg::stringValue::create("\u597d");
but it not works.
Can you tell me how to do that? thanks!

**Comment by vsg-dev** [comment]
For large .ttf the SDF generation that the vsgXchange::freetype ReaderWriter does can be quite slow, so converting to native .vsgb will be the way of avoiding this performance overhead.    I recommend using .vsgb rather than .vsgt as the binary format is small and significantly faster.

As a general note, the VisualStudio debug build is very slow so you may find using a release build might provide good enough performance without converting to native .vsgb.

When I wrote vsg::Text functionality I never had Chinese test sample to work with so just tested with what I know which is European fonts which work fine with standard ascii character set so std::stringValue works fine.  I knew at some point users would want to use extended characters sets so made it possible to assign the text as a vsg::Data leaving the door open to use std::ushortArray and std::uintArray so you could pack your strings into either of these.

There isn't any UTF8 support built into vsgText and I'm inclined to maintain this as vsg::Text supports doing GpuLayout and while we could write a vertex shader that decodes UTF8 it really would be a waste of GPU resources.  

Perhaps a better way would be for us to add a vsg::wstringValue to compliment the existing vsg::stringValue that is predefined in include/vsg/core/Value.h.  You should be able to use just use vsg::Value<std::wstring> right now but I haven;t ever tried it.

**Reply by xjgong** [reply]
My notebook(i7-9750H) to test the time with a Font(choose STXINGKA.TTF, a small size font file 3.8MB),Use VisualStudio2022 x64-Release build.
1.  vsgconv.exe will spend 80 second,the converted file vsgb has size 39MB.
2.  start vsgExamples\vsgtext(use vsgXchange to load ttf file) will spend 82 second.

**Reply by robertosfield** [reply]
The SDF computation is expensive for a large number of glyphs, I presume this font has a large number of glyphs.  Potentially one could move the computation to the GPU or do the compute multi-threaded, however, this isn't something I have to time to do for free, the focus has been get it working.

**Comment by vsg-dev** [comment]
I have done a quick test and found the the vsg::Text didn't support vsg::Value<std::wstring> so I've added a vsg::wstringValue to compliment the exiting vsg::stringValue and added support for the new wstringValue to the vsg::Text functionality.  These changes are checked into wstringValue branch of the VSG:

https://github.com/vsg-dev/VulkanSceneGraph/tree/wstringValue

I've tested it with a quick mode to vsgtext example:

            text->text = vsg::wstringValue::create(L"WIDE VulkanSceneGraph now\nhas SDF text support.");

@xjgong could you try out the wstringValue branch and let me know if that allows you to use wstringValue with the fonts and text you want to use.

**Comment by robertosfield** [comment]
In trying to add the reading/writing of std::wstring to the
vsg::Input/Output classes I have tripped up on wchat_t being 16bit under
Windows vs 32bit under Linux.  Ouch.

For vsg::AsciiInput/AsciiOutput I have already implemented a conversion of
wstring to utf8 string so perhaps doing the same for binary will be
sufficient.

**Reply by vsg-dev** [reply, marked as answer]
I have checked in read/write support for wstring with commit: 46230aa11270b5096941cdaf321f142fa63c8180

@xjgong could you test out the wstringValue branch and let me know it works for you, if there aren't any problems then I'll merge with VSG master.

**Reply by xjgong** [reply]
thanks for your answer! I will try the wstringValue.

**Reply by xjgong** [reply]
It's working!
Just load a Chinese Font file then modify the code line ( Add two Chinese characters in the end):
text->text = vsg::wstringValue::create(L"VulkanSceneGraph now\nhas SDF text support.\n 你好");
Thanks.
![vsgRenderwstring](https://github.com/vsg-dev/VulkanSceneGraph/assets/16321379/5ad994b6-c935-40d6-b056-5ef1012bcd8d)

**Reply by robertosfield** [reply]
Thanks for testing, good to hear it working well.  I have now merged the wstring branch with VSG master.
